File: go/367isPerfectSquare/367.有效的完全平方数.go
/*
 * @lc app=leetcode.cn id=367 lang=golang
 *
 * [367] 有效的完全平方数
 */

// @lc code=start
func isPerfectSquare(num int) bool {
	// 完全平方数是指 s = n*n, n 是整数
	// 解法1. 二分查找，时间复杂度 O(log n)
	// 解法2. 利用奇数数列和公式 1 + 3 + 5 + ... + (2N−1) = n*n , 时间复杂度 O(sqrt(N))
	odd := 1
	for num > 0 {
		num -= odd
		odd += 2
	}
	return num == 0
}
// @lc code=end

File: java/367valid-perfect-square/367.有效的完全平方数.java
/*
 * @lc app=leetcode.cn id=367 lang=java
 *
 * [367] 有效的完全平方数
 */

// @lc code=start
class Solution {
    public boolean isPerfectSquare(int num) {
        if (num <= 0) {
            return false;
        }
        if (num == 1) {
            return true;
        }
        // 二分法必须使用 long 类型，否则 mid 平方溢出
        long left = 2;
        long right = num / 2;
        while (left <= right) {
            long mid = left + (right - left) / 2;
            long sqrt = mid * mid;
            if (sqrt == num) {
                return true;
            } else if (sqrt > num) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return false;
    }
}
// @lc code=end

